Allow accordion and non-accordion controls on the same collapsible

nikita240 opened the pull request:
Fixes #15741.
## The setup

Suppose you have collapsible panels. You want to have dual controls for them. That means, 2 separate buttons for collapsing each panel. Now, this would normally work with bootstrap, but you want one set of controls to have [accordion](http://getbootstrap.com/javascript/#collapse-example-accordion) like behavior, and the other set of controls to have normal behavior. 

![image](https://cloud.githubusercontent.com/assets/6479817/6159059/bdfe1e14-b205-11e4-991f-5af735862e2a.png)
## The problem

The collapsible panels assume the control state of the first button you click. So if you click an accordion button first, the non-accordion button will behave like the accordion button, and vice-versa. 

[Demo](https://github.com/Nikita240/bootstrap/tree/master-test) (clone the branch and open test.html)
## The cause

Bootstrap creates a new `Collapse` object for each collapsible panel. The problem is, it creates it once during the first collapse event. This means all the options of the `Collapse` object stay the same. Because of this, when you trigger another collapse with a different controller, it still "thinks" you triggered it from the old one, because the `Collapse` objects parameters still point to the old controller. 
## The solution

Set the `$trigger` property and recalculate parent during each controller click.

[Demo](https://github.com/Nikita240/bootstrap/tree/collapsible-accordion-multiple-controls-test) (clone the branch and open test.html)

nikita240 committed:
Re-set collapse trigger on every click

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 60debf2c8f45e182febdf69cad5f7183af45fe9a
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50424782

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
Please add a new relevant unit test to https://github.com/twbs/bootstrap/blob/master/js/tests/unit/collapse.js

nikita240 committed:
Add collapse.js unit test for 60debf2

Nikita240 commented:
@cvrebert Done f515a0e637fd0311bd6ac413214c22c45d637cb6

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: f515a0e637fd0311bd6ac413214c22c45d637cb6
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50437054

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Nikita240 commented:
You know what, don't actually accept this pull request yet.

Look into it, when you have multiple collapsible triggers, it seems the `aria-expanded` and `collapsed` class doesn't propagate properly between. It probably was never bootstraps intention to support multiple triggers, but I personally don't see a reason why not.

I'll modify the code further to support multiple triggers.

nikita240 committed:
Propagate collapsed state to multiple triggers

Upon construction of a `Collapse` object, find all collapsible triggers
that have the same target as the clicked trigger.

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 1c56c72074f21a90714f2796905bc28581035f53
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50562174

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

nikita240 committed:
Cleanup uneccessary function in collapse.js

nikita240 committed:
Add collapse.js unit test for 1c56c72

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 3b2a80a9f138da12260c87654616ebc9f84a2d66
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50565277

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Nikita240 commented:
Pull request is ready.

`aria-expanded` and `collapsed` class now propagate to all triggers that point to the same target.

Also added a unit test for the propagation.

Nikita240 commented:
Oh. This also resolves #15741.

nikita240 committed:
Remove uneccessary `.find()`

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 630b5f876a7069e2b7431a8a268dcf5e6d39f954
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50567117

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented on js/collapse.js:
```diff
@@ -16,14 +16,10 @@
     this.$element      = $(element)
     this.options       = $.extend({}, Collapse.DEFAULTS, options)
-    this.$trigger      = $(this.options.trigger).filter('[href="#' + element.id + '"], [data-target="#' + element.id + '"]')
+    this.$allTriggers  = $('[data-toggle="collapse"]').filter('[href="#' + this.$element[0].id + '"], [data-target="#' + this.$element[0].id + '"]')
```
Perhaps we should compute this on-demand, so that it doesn't become outdated when folks add/delete triggers after the Collapse gets initialized?

Nikita240 replied:
I was thinking about that too. Good idea.

Nikita240 replied:
Done

nikita240 committed:
Replace `.filter()` with a more complex selector

cvrebert commented on js/collapse.js:
```diff
@@ -154,6 +150,19 @@
   }
 
+  Collapse.prototype.setCurrentTrigger = function ($trigger) {
+    this.options.trigger = $trigger;
```
remove semicolon to comply with project's coding style

Nikita240 replied:
Done

Nikita240 commented:
@cvrebert The `.filter()` function is identical to the one in the original code. 

I added the commit, but in my opinion the `.filter()` function makes the code cleaner, as it reduces the need of repeating the `[data-toggle="collapse"]` of the selector.

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 48f1657580783439d30b61e57157d1a41d659515
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50568748

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
Competing PR: #15751.

nikita240 committed:
Remove accidental semicolon

nikita240 committed:
Compute `$allTriggers` on every trigger click

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: 00e47ad34dc50ba5256123247317c37820fee89c
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/50569760

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

cvrebert commented:
@Nikita240 Sorry for the trouble, but would you mind merging in or rebasing against `master` to resolve the merge conflicts? Thanks, much appreciated.

nikita240 committed:
Merge branch 'master' into collapsible-accordion-multiple-controls

Conflicts:
	dist/js/bootstrap.js
	dist/js/bootstrap.min.js
	docs/assets/js/customize.min.js
	docs/assets/js/raw-files.min.js
	docs/dist/js/bootstrap.js
	docs/dist/js/bootstrap.min.js
	js/collapse.js

twbs-savage commented:
**Tests passed.** Automated cross-browser testing via Sauce Labs and Travis CI shows that the JavaScript changes in this pull request are: **CONFIRMED**

Commit: cc48197eb16069987621fe94fa16b7a1b678b089
Build details: https://travis-ci.org/twbs-savage/bootstrap/builds/54631334

(*Please note that this is a [fully automated](https://github.com/twbs/savage) comment.*)

Nikita240 commented:
@cvrebert Done! Took me a while to figure out all the unit test changes.

cvrebert commented on js/collapse.js:
```diff
@@ -154,6 +149,21 @@
   }
 
+  Collapse.prototype.updateTriggers = function ($currentTrigger) {
+    this.$allTriggers  = $('[data-toggle="collapse"][href="#' + this.$element[0].id + '"],'
```
Should these also use `options.trigger` instead of hardcoding `[data-toggle="collapse"]` ?

Nikita240 replied:
No, because `[data-toggle="collapse"]` is used as an identifier to find all triggers. `options.trigger` is set to point to the actual trigger that triggered the collapse.

cvrebert commented on js/tests/unit/collapse.js:
```diff
@@ -440,4 +440,158 @@ $(function () {
+        + '</div>'
+    var $groups = $(accordionHTML).appendTo('#qunit-fixture').find('.panel')
+
+    var $body1 = $('<div id="body1" aria-expanded="true" class="in"/>').appendTo($groups.eq(0))
```
Perhaps move this HTML into `accordionHTML` and then just reference it here via `$('#body1')` ?
Ditto for body2 and body3.

Nikita240 replied:
I can of course change it if you want, but I was following the style of the other unit tests:

```javascript
  QUnit.test('should not fire show event if show is prevented because other element is still transitioning', function (assert) {
    assert.expect(1)
    var done = assert.async()

    var accordionHTML = '<div id="accordion">'
        + '<div class="panel"/>'
        + '<div class="panel"/>'
        + '</div>'
    var showFired = false
    var $groups   = $(accordionHTML).appendTo('#qunit-fixture').find('.panel')

    var $target1 = $('<a data-toggle="collapse" href="#body1" data-parent="#accordion"/>').appendTo($groups.eq(0))

    $('<div id="body1" class="collapse"/>')
      .appendTo($groups.eq(0))
      .on('show.bs.collapse', function () {
        showFired = true
      })

    var $target2 = $('<a data-toggle="collapse" href="#body2" data-parent="#accordion"/>').appendTo($groups.eq(1))
    var $body2   = $('<div id="body2" class="collapse"/>').appendTo($groups.eq(1))

    $target2.trigger('click')

    $body2
      .toggleClass('in collapsing')
      .data('bs.collapse').transitioning = 1

    $target1.trigger('click')

    setTimeout(function () {
      assert.ok(!showFired, 'show event did not fire')
      done()
    }, 1)
  })
```

cvrebert replied:
Fair enough

mdo commented:
I understand the desired behavior, but what's the use case? Are there practical examples that this really addresses? I don't think we'd want to change the behavior of the component based on what trigger was used.

mdo commented:
Punting on this for now regardless of a response to the above given it's a new feature and we're no longer adding those to v4.

mdo commented:
Please holler with any and all feedback though—don't let me closing it dissuade you completely :).
